File: src/main/java/com/extendedae_plus/client/render/widgets/button/EAEPServerCycleButton.java
package com.extendedae_plus.client.render.widgets.button;

import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.neoforged.neoforge.network.PacketDistributor;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class EAEPServerCycleButton extends EAEPCycleButton {
    private final Supplier<Integer> syncerState;

    public EAEPServerCycleButton(List<EAEPActionItems> states,
                                 Consumer<EAEPActionItems> taskSingle,
                                 @Nullable IteratorState iteratorState,
                                 Supplier<Integer> syncerState) {
        super(states, (index, action) -> taskSingle.accept(action), iteratorState);
        this.syncerState = syncerState;
    }

    @Override
    public int iterateState(boolean reversed) {
        super.iterateState(reversed);
        if (this.stateIndex != this.syncerState.get())
            this.setStateIndex(this.syncerState.get());
        return this.stateIndex;
    }

    public void updateState() {
        this.setStateIndex(this.syncerState.get());
    }

    public static final class Builder {
        private final List<EAEPActionItems> states = new ArrayList<>();
        private Consumer<EAEPActionItems> task = ignored -> {};
        private IteratorState iteratorState = null;
        private Supplier<Integer> syncerState = () -> 0;

        public Builder setTask(CustomPacketPayload task) {
            return this.setTask(ignored -> PacketDistributor.sendToServer(task));
        }

        public Builder setTask(Runnable task) {
            return this.setTask(ignored -> task.run());
        }

        public Builder setTask(Consumer<EAEPActionItems> task) {
            this.task = task;
            return this;
        }

        public Builder addPart(EAEPActionItems action) {
            this.states.add(action);
            return this;
        }

        public Builder setIterator(IteratorState iteratorState) {
            this.iteratorState = iteratorState;
            return this;
        }

        public Builder setSyncer(Supplier<Integer> syncer) {
            this.syncerState = syncer;
            return this;
        }

        public Builder setSyncer(SyncerBooleanGeneric syncer) {
            this.syncerState = syncer;
            return this;
        }

        public <TEnum extends Enum<TEnum>> Builder setSyncer(SyncerEnumGeneric<TEnum> syncer) {
            this.syncerState = syncer;
            return this;
        }

        public EAEPServerCycleButton build() {
            return new EAEPServerCycleButton(this.states, this.task, this.iteratorState, this.syncerState);
        }
    }

    @FunctionalInterface
    public interface SyncerBooleanGeneric extends Supplier<Integer> {
        boolean getState();

        default Integer get() {
            return this.getState() ? 1 : 0;
        }
    }

    @FunctionalInterface
    public interface SyncerEnumGeneric<TEnum extends Enum<TEnum>> extends Supplier<Integer> {
        TEnum getState();

        default Integer get() {
            return this.getState().ordinal();
        }
    }
}

File: src/main/kotlin/com/extendedae_plus/client/render/widgets/button/EAEPServerCycleButton.kt
package com.extendedae_plus.client.render.widgets.button

import com.fish.fishlib.network.base.PacketGeneric.Companion.sendToServer
import net.minecraft.network.protocol.common.custom.CustomPacketPayload

class EAEPServerCycleButton(
    states: List<EAEPActionItems>,
    taskSingle: (EAEPActionItems) -> Unit,
    iteratorState: IteratorState?,
    private val syncerState: () -> Int
) : EAEPCycleButton(
    states,
    { _, action -> taskSingle(action) },
    iteratorState
) {
    override fun iterateState(reversed: Boolean): Int {
        super.iterateState(reversed)
        val synced = this.syncerState()
        if (this.indexState != synced)
            this.indexState = synced
        return this.indexState
    }

    fun updateState() = this.setStateIndex(this.syncerState())

    class Builder {
        private val states = ArrayList<EAEPActionItems>()
        private var task = { _: EAEPActionItems -> }
        private var iteratorState: IteratorState? = null
        private var syncerState: (() -> Int)? = null

        fun setTask(task: CustomPacketPayload) =
            this.setTask { _ -> task.sendToServer() }

        fun setTask(task: Runnable) =
            this.setTask { _ -> task.run() }

        fun setTask(task: (EAEPActionItems) -> Unit): Builder {
            this.task = task
            return this
        }

        fun addPart(action: EAEPActionItems): Builder {
            this.states += action
            return this
        }

        fun setIterator(iteratorState: IteratorState?): Builder {
            this.iteratorState = iteratorState
            return this
        }

        fun setSyncer(syncer: () -> Int): Builder {
            this.syncerState = syncer
            return this
        }

//        fun setSyncer(syncer: SyncerBooleanGeneric): Builder {
//            this.syncerState = syncer
//            return this
//        }
//
//        fun <TEnum : Enum<TEnum>> setSyncer(syncer: SyncerEnumGeneric<TEnum>): Builder {
//            this.syncerState = syncer
//            return this
//        }

        fun build() = EAEPServerCycleButton(
            this.states,
            this.task,
            this.iteratorState,
            this.syncerState
                ?: throw IllegalArgumentException("Cannot build button without setSyncer call")
        )
    }

}

fun interface SyncerEnumGeneric<TEnum : Enum<TEnum>> : (() -> Int) {
    fun get(): TEnum

    override operator fun invoke() = this.get().ordinal
}

fun interface SyncerBooleanGeneric : (() -> Int) {
    fun get(): Boolean

    override operator fun invoke() = if (this.get()) 1 else 0
}